### ISSUE TITLE
Quadratic (?) increase when rbinding nested data frames

### ISSUE THREAD

**Issue by hadley** [r-lib/vctrs#530]
``` r
library(vctrs)

df <- data.frame(x = 1)
df$y <- df
df$x <- NULL

x <- list(df)

x2 <- rep(x, 1e2)
x3 <- rep(x, 1e3)
x4 <- rep(x, 1e4)

bench::mark(
  vec_rbind(!!!x2),
  vec_rbind(!!!x3),
  vec_rbind(!!!x4),
  check = FALSE
)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 3 x 6
#>   expression            min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>       <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_rbind(!!!x2)   1.07ms   1.21ms   763.       50.8KB    16.0 
#> 2 vec_rbind(!!!x3)  15.93ms  19.11ms    51.6      3.88MB     9.93
#> 3 vec_rbind(!!!x4)    1.12s    1.12s     0.896  382.12MB     8.96
```

<sup>Created on 2019-08-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Doesn't happen with regular columns. Part of problem in https://github.com/tidyverse/tidyr/issues/694

**Comment by DavisVaughan** [comment]
Looks like this might have to do with `df_assign()` and its call to `vec_restore()`? Maybe we should take a recursive proxy of `out` before doing the assignment loop in `vec_rbind()`, and then do a recursive `vec_restore()` after the assignment loop. It would require modifying `df_assign()` to assume that you have done a recursive proxy already

**Comment by DavisVaughan** [comment]
So, good news! I think this was fixed in #537.

I'm pretty sure the problem was in the spot in `df_assign()` I mentioned above, where `vec_restore()` was repeatedly fully materializing automatic row names.

This means that tidyverse/tidyr#751 is _greatly_ improved because of this, and a few extra seconds are also gained there from #536.

With dev vctrs, and going up to `1e5`:

``` r
library(vctrs)

df <- data.frame(x = 1)
df$y <- df
df$x <- NULL

x <- list(df)

x2 <- rep(x, 1e2)
x3 <- rep(x, 1e3)
x4 <- rep(x, 1e4)
x5 <- rep(x, 1e5)

bench::mark(
  vec_rbind(!!!x2),
  vec_rbind(!!!x3),
  vec_rbind(!!!x4),
  vec_rbind(!!!x5),
  check = FALSE,
  min_iterations = 10
)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 4 x 6
#>   expression            min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>       <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_rbind(!!!x2)  731.5µs  923.1µs   740.       7.07KB     4.00
#> 2 vec_rbind(!!!x3)   7.56ms   8.11ms   120.      19.67KB     3.99
#> 3 vec_rbind(!!!x4)  80.16ms  87.08ms    10.6    195.45KB     5.30
#> 4 vec_rbind(!!!x5)    1.23s    1.38s     0.702    1.91MB     6.53
```

<sup>Created on 2019-09-24 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>
